core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,749 @@
+        } catch (Throwable t) {
+          LOG.log(Level.INFO, "Exception notifying context listener", t);
+        }
+      }
+      parent.removeListener(parentListener);
+    }
+  }
+
+  // Used in tests to ensure that listeners are defined and released based on
+  // cancellation propagation. It's very important to ensure that we do not
+  // accidentally retain listeners.
+  int listenerCount() {
+    synchronized (this) {
+      return listeners == null ? 0 : listeners.size();
+    }
+  }
+
+  /**
+   * Wrap a {@link Runnable} so that it executes with this context as the {@link #current} context.
+   */
+  public Runnable wrap(final Runnable r) {
+    return new Runnable() {
+      @Override
+      public void run() {
+        attach();
+        try {
+          r.run();
+        } finally {
+          detach();
+        }
+      }
+    };
+  }
+
+  /**
+   * Wrap a {@link Callable} so that it executes with this context as the {@link #current} context.
+   */
+  public <C> Callable<C> wrap(final Callable<C> c) {
+    return new Callable<C>() {
+      @Override
+      public C call() throws Exception {
+        attach();
+        try {
+          return c.call();
+        } finally {
+          detach();
+        }
+      }
+    };
+  }
+
+  /**
+   * Lookup the value for a key in the context inheritance chain.
+   */
+  private Object lookup(Key key) {
+    for (int i = 0; i < keyValueEntries.length; i++) {
+      if (key.equals(keyValueEntries[i][0])) {
+        return keyValueEntries[i][1];
+      }
+    }
+    if (parent == null) {
+      return null;
+    }
+    return parent.lookup(key);
+  }
+
+  /**
+   * A context which inherits cancellation from its parent but which can also be independently
+   * cancelled and which will propagate cancellation to its descendants.
+   */
+  public static final class CancellableContext extends Context {
+
+    /**
+     * Used with cancelledUpdater. 0->not cancelled, 1->cancelled.
+     */
+    @SuppressWarnings("unused")
+    private boolean cancelled;
+    private Throwable cause;
+    private final Context dummy;
+    private ScheduledFuture<?> scheduledFuture;
+
+    /**
+     * Create a cancellable context that does not have a deadline.
+     */
+    private CancellableContext(Context parent) {
+      super(parent, EMPTY_ENTRIES);
+      // Create a dummy that inherits from this to attach and detach so that you cannot retrieve a
+      // cancellable context from Context.current()
+      dummy = new Context(this, EMPTY_ENTRIES);
+    }
+
+    /**
+     * Create a cancellable context that has a deadline.
+     */
+    private CancellableContext(Context parent, long delayNanos) {
+      super(parent, EMPTY_ENTRIES);
+      // Create a dummy that inherits from this to attach and detach so that you cannot retrieve a
+      // cancellable context from Context.current()
+      dummy = new Context(this, EMPTY_ENTRIES);
+      scheduledFuture = SCHEDULER.create().schedule(new Runnable() {
```

Comment:
This doesn't share executors at all. You need to call `SharedResourceHolder.get(SCHEDULER)` here, and then later call `SharedResourceHolder.release(SCHEDULER)` (probably within the Runnable here).